api/challenge.go

```diff
@@ -0,0 +1,95 @@
+package api
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/ImageWare/TLSential/challenge_config"
+	"github.com/ImageWare/TLSential/model"
+)
+
+var (
+	ErrInvalidAuthEmail = errors.New("auth email cannot be blank")
+	ErrInvalidAuthKey   = errors.New("auth key cannot be blank")
+)
+
+// ChallengeHandler provides endpoints for all api/config/ calls.
+// TODO: Change this to a "Handle" func only, abstract from there.
+type ChallengeHandler interface {
+	Get() http.HandlerFunc
+	Put() http.HandlerFunc
+}
+
+type challengeHandler struct {
+	cs challenge_config.Service
+}
+
+// NewChallengeHandler takes a config.Service and returns a working ConfigHandler.
+func NewChallengeHandler(cs challenge_config.Service) ChallengeHandler {
+	return &challengeHandler{cs}
+}
+
+// Get responds to the api/challenge endpoint.
+func (h *challengeHandler) Get() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		c, err := h.cs.Auth()
+		if err != nil {
+			log.Printf("challengeHandlerGET, Auth(), %s", err.Error())
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		out, err := json.Marshal(c)
+		if err != nil {
+			log.Printf("challengeHandlerGET, json.Marshal(), %s", err.Error())
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintf(w, "%s", out)
+	}
+}
+
+func (h *challengeHandler) Put() http.HandlerFunc {
```

Comment:
Just a nit but I think it could be helpful to have like

// Put /api/config 

above these functions just as a convenience thing

Reply:
also maybe add the permission levels etc
I know these can be gathered from reading the code but I think it makes it a little easier to grok at a glance

Reply:
Good ideas